src/es6-module-utils.ts

```diff
@@ -0,0 +1,444 @@
+ * http://polymer.github.io/PATENTS.txt
+ */
+import traverse, {NodePath} from 'babel-traverse';
+import * as babel from 'babel-types';
+import * as clone from 'clone';
+import {Analyzer, FileRelativeUrl, PackageRelativeUrl, ResolvedUrl} from 'polymer-analyzer';
+import {rollup} from 'rollup';
+
+import {getAnalysisDocument} from './analyzer-utils';
+import {getNodePath, getNodeValue, parseModuleFile, rewriteNode, serialize} from './babel-utils';
+import {AssignedBundle, BundleManifest} from './bundle-manifest';
+import {Bundler} from './bundler';
+import {ensureLeadingDot, getFileName} from './url-utils';
+import {camelCase} from './utils';
+
+/**
+ * Looks up and/or defines the unique name for an item exported with the given
+ * name in a module within a in a bundle.
+ */
+export function getBundleModuleExportName(
+    bundle: AssignedBundle, moduleUrl: ResolvedUrl, name: string): string {
+  let basisBundle = false;
+  // When the bundle contains a file with same URL as the bundle, then that
+  // module's exports will remain unchanged.
+  if (bundle.bundle.files.has(bundle.url) && moduleUrl === bundle.url) {
+    //    basisBundle = true;
+  }
+  let moduleExports = bundle.bundle.bundledExports.get(moduleUrl);
+  const bundledExports = bundle.bundle.bundledExports;
+  if (!moduleExports) {
+    moduleExports = new Map<string, string>();
+    bundledExports.set(moduleUrl, moduleExports);
+  }
+  let exportName = moduleExports.get(name);
+  if (!exportName) {
+    let trialName = name;
+    if (!basisBundle) {
+      let moduleFileNameIdentifier =
+          '$' + camelCase(getFileName(moduleUrl).replace(/\.[a-z]+$/, ''));
+      trialName =
+          trialName.replace(/^default$/, `${moduleFileNameIdentifier}Default`)
+              .replace(/^\*$/, moduleFileNameIdentifier)
+              .replace(/[^a-z0-9_]/gi, '$');
+    }
+    while (!exportName) {
+      if ([...bundledExports.values()].every(
+              (map) => [...map.values()].indexOf(trialName) === -1)) {
+        exportName = trialName;
+      } else {
+        if (trialName.match(/\$[0-9]+$/)) {
+          trialName = trialName.replace(/[0-9]+$/, (v) => `${parseInt(v) + 1}`);
+        } else {
+          trialName = `${trialName}$1`;
+        }
+      }
+    }
+    moduleExports.set(name, exportName);
+  }
+  return exportName;
+}
+
+export function hasDefaultModuleExport(node: babel.Node): boolean {
+  let hasDefaultModuleExport = false;
+  traverse(node, {
+    noScope: true,
+    ExportDefaultDeclaration: {
+      enter(path: NodePath) {
+        hasDefaultModuleExport = true;
+        path.stop();
+      }
+    }
+  });
+  return hasDefaultModuleExport;
+}
+
+export function getModuleExportNames(node: babel.Node): Set<string> {
+  const exportedNames: string[] = [];
+  traverse(node, {
+    noScope: true,
+    ExportNamedDeclaration: {
+      enter(path: NodePath) {
+        const exportNode: babel.ExportNamedDeclaration = path.node as any;
```

Comment:
why the any? (comment)

Reply:
Oh this was silly.  Changed to just `const exportNode = path.node as babel.ExportNamedDeclaration;` in commit 066031b79fbf4181080892d8cbb7955478c34de2